runtime/tr.source/trj9/optimizer/J9TransformUtil.cpp

```diff
@@ -264,43 +317,92 @@ static void *dereferenceStructPointerChain(void *baseStruct, TR::Node *baseNode,
    //    - Verifiable if it is known to point to a struct of the proper type or null
    //    - Verified trivially by virtue of being verifiable and non null
    //
-   // The onus is on the caller to ensure that baseStruct/baseNode is
-   // verifiable.  This is why we have separated the notion of "verifiable"
-   // from "verified": there is less burden on the caller if it only needs to
-   // ensure verifiability.
```

Comment:
It was accurate before the change, where the user passes in the address child of the node to be dereferenced. With this change, the user will pass in the node to be dereferenced and we'll verify field access before the dereference, i.e. move the verification from caller to `dereferenceStructPointerChain`. The number of verifications before and after this change remains the same.

Reply:
Doesn't `baseStruct` still have to be verifiable?

Reply:
I believe verifiability check only applies to a field access node with symref. According to the definition of verifiability, the field has to be known to point at something of proper type or null. The verifiability check is done by checking the symref on the field node. Access denoted by `baseNode` doesn't have to be verifiable due to:
1. `baseNode` can be a direct load and its symref can be a temp which tells nothing about the access. Even worse, if `baseNode` is `aconst`, there is no symref. 
2.  The caller can give us any value for `baseStruct/baseNode` as long as it thinks `baseNode` has value `baseStruct`. This happens a lot in VP where VP propagates constraints to nodes which turns some field accesses dereferencible at compile time.
3. We currently only verify field accesses which are indirect loads due to above reasons.

Although the caller can give us any value for `baseStruct/baseNode`, `dereferenceStructPointerChain` is written in a way that no input can crash it. If it was given a garbage, it'll return a garbage (most likely to happen when dereferencing a native struct) or `NULL` when the access cannot be verified or dereferenced.

Reply:
Thanks Devin for pointing out something I missed. Here's my understanding after the discussion.
Verifiability is a property on the data to be looked at and the access to be performed on the data.
The type of the data has to match the access, which is represented by IL trees to qualify for verifiability.

baseNode/baseStruct is verifiable if 
1) The trees indicates baseNode points to a java object or null; baseStruct points to a java object or null.
2) The trees indicates baseNode is a native struct of type A or null; baseStruct points to a native struct of type A or null.

It's ok to have baseNode and baseStruct pointing to two different types of java objects since the verification step can do dynamic type check. This way, we can tolerate mismatches between the data and the trees when we're looking at dead code on impossible pathes. However, dereferencing from data that is not verifiable might still crash the JIT, if not, garbage will be returned.

Thus the onus on the caller to pass in something that's verifiable still applies. I'll put the comment back and move them to `transformIndirectLoadChain/transformIndirectLoadChainAt` so callers are more aware of their responsibilities.